src/hazelcore/UHExp.re

```diff
@@ -305,10 +305,119 @@ let text_operand =
     );
   };
 
-let associate = (seq: seq) => {
-  let skel_str = Skel.mk_skel_str(seq, Operators_Exp.to_parse_string);
-  let lexbuf = Lexing.from_string(skel_str);
-  SkelExprParser.skel_expr(SkelExprLexer.read, lexbuf);
+let associate = (seq: seq): skel => {
+  /**
+   * Shunting-yard opp algorithm according to the specification here
+   * https://en.wikipedia.org/wiki/Shunting-yard_algorithm#The_algorithm_in_detail
+   *
+   */
+  let rec go_seq =
+          (
+            skels: list(skel), /* List of skels to be combined into single output skel. */
+            op_stack: list(Operators_Exp.t), /* Stack of operators. */
+            seq: seq, /* Convert this seq to output skel. */
+            lex_addr: int,
+          ) /* Lexical address of the current operand. */
+          : skel => {
+    switch (seq) {
+    | S(_, affix) =>
+      /**
+       * If the next token is an operand, add a new operand to the output skel.
+       */
+      go_affix(
+        [Skel.Placeholder(lex_addr), ...skels],
+        op_stack,
+        affix, /* Tail of seq without first operand. */
+        lex_addr + 1 /* Increment lexical address of operand in seq. */
+      )
+    };
+  }
+  and go_affix =
+      (
+        skels: list(skel),
+        op_stack: list(Operators_Exp.t),
+        affix: affix,
+        lex_addr: int,
+      )
+      : skel => {
+    switch (affix) {
+    | A(op, seq) =>
+      /**
+       * If the next token is an operator, pop operators from the operator stack,
+       * then push this operator to the operator stack.
+       */
+      let should_mv = op' =>
+        /**
+         * Continue popping operators while the precedence of the top of the operator
+         * stack has greater precedence than the current operator.
+         */
+        {
+          switch (Operators_Exp.associativity(op')) {
+          | Left =>
+            Operators_Exp.precedence(op) <= Operators_Exp.precedence(op')
+          | Right =>
+            Operators_Exp.precedence(op) < Operators_Exp.precedence(op')
+          };
+        };
+
+      let (skels', op_stack') = build_ast_while(skels, op_stack, should_mv);
+
+      /* Push this operator to the operator stack. */
+      go_seq(skels', [op, ...op_stack'], seq, lex_addr);
+    | E =>
+      /**
+       * Once the input seq is empty, continuously pop
+       * operators in the stack and build up the output skel.
+       */
+      let (skels', _) = build_ast_while(skels, op_stack, _ => true);
+      List.hd(skels');
+    };
+  }
+  and build_ast_while =
+      (skels: list(skel), op_stack: list(Operators_Exp.t), should_mv)
+      : (list(skel), list(Operators_Exp.t)) => {
+    /* Move operators from the operator stack to the output skel list while... */
+    switch (op_stack) {
+    | [] => (skels, op_stack) /* (1) The operator stack is not empty. */
+    | [op, ...op_stack'] =>
+      if (should_mv(op)) {
+        /**  (2) The operator on the top of the stack has greater precedence than the current.
```

Comment:
don't you also have to consider the situation where the operator at the top of the operator stack has equal precedence and the token is left associative?

Reply:
I believe I am doing that inside the [should_mv predicate](https://github.com/hazelgrove/hazel/pull/361/files/f50d0d84ba032b8fda1cd59b875351e781d7aa9a#diff-07032a218ad58ee406b8c5f9f3b5cb41R349-R361) that I pass into this function.